File: backend/src/main/java/com/tjh/riskfactor/entity/form/Question.java
package com.tjh.riskfactor.entity.form;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.experimental.Accessors;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import static com.fasterxml.jackson.annotation.JsonInclude.Include;

import static com.tjh.riskfactor.util.Utils.annotatedDeclaredFields;

import javax.persistence.*;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Data @Entity
@Table(name = "question")
@Accessors(chain = true)
@EqualsAndHashCode(of = "id")
public class Question {

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    private QuestionType type;

    private String label;

    @ManyToMany(cascade = CascadeType.REMOVE)
    @JoinTable(name = "question_list",
        joinColumns = @JoinColumn(name = "head", referencedColumnName = "id"),
        inverseJoinColumns = @JoinColumn(name = "qid", referencedColumnName = "id")
    )
    @OrderColumn(name = "sequence", nullable = false)
    @JsonInclude(Include.NON_EMPTY)
    private List<Question> list;

    // 指定是否为必填
    @Transient private Boolean required;
    // 指定该问题的输入能否控制后续问题的显示
    // 用于指定联动问题，即只有这个问题输入之后才能显示后续问题
    @Transient private Boolean isEnabler;
    // 问题标签位置，仅供输入类组件使用，表示标签的位置
    // 值的含义同addonPosition
    @Transient private String labelPosition;
    // 指定YESNO_CHOICE问题的选项内容
    // 格式：[是含义]/[否含义]
    @Transient private String yesno;
    // 指定该问题是问题组长的前置/后置输入控件
    //   比如 [输入持续时间（整数）][单位选择（天/月）]，后者是前者的后置输入控件
    // 该域的含义：
    //   null: 无效果（默认）
    //   prefix: 前置输入
    //   postfix: 后置输入
    @Transient private String addonPosition;
    // 指定列表/多选类问题默认选项，默认是未选择
    // 对于单项选择来说，只有一个条目
    // 对于多项选择来说，有多个条目，由半角逗号分隔
    @Transient private String selected;
    // 额外说明
    @Transient private String description;
    // 占位文字，一般是输入类控件需要
    // 对于IMMUTABLE类型来说，则是填充字符串
    @Transient private String placeholder;
    // 下拉菜单如果选项很多，那么需要有输入时筛选的功能
    // 指定输入时筛选的前缀key，仅用于CHOICE
    // 对于下拉选择类问题来说，只要这一项不为空则开启输入筛选功能
    @Transient private String filterKey;

    @Column(name = "option", length = 512)
    @Access(AccessType.PROPERTY)
    @JsonIgnore
    public String getOption() {
        String fields = annotatedDeclaredFields(this.getClass(), Transient.class).map(field -> {
            field.setAccessible(true);
            var name = field.getName();
            try {
                var value = field.get(this);
                if(value == null)
                    return null;
                if(field.getType().equals(Boolean.class) && ((Boolean) value))
                    return name;
                else
                    return String.format("%s:%s", name, value);
            } catch(IllegalAccessException ignored) { return null; }
        }).filter(Objects::nonNull).collect(Collectors.joining("$;"));

        return fields.isEmpty() ? null : fields;
    }

    public void setOption(String option) {
        if(option == null)
            return;
        Arrays.stream(option.split("\\$;")).forEach(part -> {
            var value = part.split(":");
            try {
                var field = this.getClass().getDeclaredField(value[0]);
                field.setAccessible(true);
                if(value.length == 1)
                    field.set(this, true);
                else
                    field.set(this, value[1]);
            } catch (IllegalAccessException | NoSuchFieldException ignored) {}
        });
    }

}

File: backend/src/main/kotlin/com/tjh/riskfactor/entity/form/Question.kt
package com.tjh.riskfactor.entity.form

import com.fasterxml.jackson.annotation.*

import com.tjh.riskfactor.entity.IEntity
import com.tjh.riskfactor.entity.cvt.QuestionOptionsConverter

import javax.persistence.*

@Entity @Table(name = "question")
class Question(
    /**
     * 可为null，代表是辅助用的纯文本
     */
    @get:Column
    @get:Enumerated
    var type: QuestionType? = null,

    @get:Column
    var label: String? = null,

    @get:ManyToMany(fetch = FetchType.EAGER, cascade = [CascadeType.REMOVE])
    @get:JoinTable(name = "question_list",
        joinColumns = [JoinColumn(name = "head", referencedColumnName = "id")],
        inverseJoinColumns = [JoinColumn(name = "qid", referencedColumnName = "id")]
    )
    @get:OrderColumn(name = "sequence", nullable = false)
    @get:JsonInclude(JsonInclude.Include.NON_EMPTY)
    var list: MutableList<Question>? = null,

    @get:Transient
    @get:JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    var ref: String? = null
): IEntity() {
    @get:Column(length = 102400)
    @get:Convert(converter = QuestionOptionsConverter::class)
    @get:JsonUnwrapped
    var options: QuestionOptions? = null
}

@Suppress("unused")
enum class QuestionType(@get:JsonValue val value: String) {
    TEXT("text"), // 对应文本输入
    NUMBER("number"), // 对应数字输入，浮点
    DATE("date"), // 对应日期选择
    IMMUTABLE("disabled"), // 不可变内容，其值由服务器赋予

    YESNO_CHOICE("either"), // 两项单选，否/是，一般选“是”有后置问题

    // 问题组类型
    LIST("list"), // 固定数量问题组
    LIST_APPENDABLE("template"), // 可变长度问题组，一般是可重复添加的项目
    // 该类问题下所属question list的第一个（且应仅有一个）为添加问题模板
    MULTI_CHOICE("choice-multi"), // 多选，一般是Checkbox实现
    SINGLE_CHOICE("choice"), // 单选，一般是RadioButton实现
    SINGLE_SELECT("select"), // 单项下拉选择
    MULTI_SELECT("select-multi"), // 多项下拉选择

    TABLE("table"),
    TABLE_HEADER("table-header"),

    HEADER("header");
}

@Suppress("unused")
@JsonInclude(JsonInclude.Include.NON_EMPTY)
open class QuestionOptions {
    /**
     * 指定是否为必填
     */
    open var required: Boolean? = null

    /**
     * 指定该问题的输入能否控制后续问题的显示。用于指定联动问题，即只有这个问题输入之后才能显示后续问题
     */
    open var isEnabler: Boolean? = null

    /**
     * 指定多选/下拉类控件的最后是否有一个“其他”用于输入自定义选项内容
     */
    open var customizable: Boolean? = null

    /**
     * 指定多选/下拉类控件的选项
     *
     * 如果需要给选项指定筛选键，用 &#91;内容&#93;/&#91;筛选键&#93; 格式
     */
    open var choices: List<String>? = null

    /**
     * 问题标签位置，仅供输入类组件使用，表示标签的位置
     * 值的含义同[addonPosition]
     */
    open var labelPosition: String? = null

    /**
     * 指定[QuestionType.YESNO_CHOICE]问题的选项内容
     *
     * 格式：&#91;是含义]/&#91;否含义]
     */
    open var yesno: String? = null

    /**
     * 指定该问题是问题组长的前置/后置输入控件，比如 &#91;输入持续时间（整数）]&#91;单位选择（天/月）]，后者是前者的后置输入控件
     *
     * 该域的含义：
     *   null: 无效果（默认）
     *   prefix: 前置输入
     *   postfix: 后置输入
     */
    open var addonPosition: String? = null

    /**
     * 指定列表/多选类问题默认选项，默认是未选择
     *
     * 对于单项选择来说，只有一个条目；对于多项选择来说，有多个条目，由半角逗号分隔
     */
    open var selected: String? = null

    /**
     * 额外说明
     */
    open var description: String? = null

    /**
     * 占位文字，一般是输入类控件需要
     *
     * 对于[QuestionType.IMMUTABLE]类型来说，则是填充字符串
     */
    open var placeholder: String? = null
}
